File: lib/features/task/handler/task_state_manager.dart
import 'dart:io';
import 'package:path/path.dart' as p;

class TaskStateManager {
  // Initial Configuration Parameters
  final String videoUrl;
  final String projectRoot; // Added from Python
  final String videoId;
  final String taskId; // Added from Python

  final String savePath = "temp/"; // Default save path - consider if still needed or replaced by projectRoot logic
  final String videoFormat = "mp4"; // Default video format
  final String audioFormat = "mp3"; // Default audio format

  // Dynamic State Parameters (evolve during processing)
  String? processedVideoPath; // This might be equivalent to outVideoPath or translateMp4
  String? extractedFramesPath;
  List<String>? extractedImageFiles;
  String? extractedAudioPath; // This might be equivalent to originalMp3
  String? extractedAudioFile;

  // Path properties from Python
  String currentDir;
  String? audioDir;
  String inputVideoPath;
  String? outVideoPath;
  String? translateMp4;
  String originalMp3;
  String? originalSrt;
  String? originalVtt;
  String? translateSrt;
  String? translateTxt;
  String? translateVtt;
  String imageCover;
  String? upVideoPath;
  String? videoCosUrl; // Note: Python version depends on 'settings' and 'video.cosKey'
  String? mindmapJson;
  String? translateMd;
  String? okFile;

  
  Map<String, dynamic> cache = {};

  // Error State
  bool hasError = false;
  String? errorMessage;
  dynamic errorDetails;

  TaskStateManager({
    required this.videoUrl,
    required this.projectRoot,

    // String? initialVideoCosUrl, // Optional: if you want to set it externally
  })  : videoId = _extractVideoId(videoUrl),
        currentDir = p.join(projectRoot, _extractVideoId(videoUrl)),
        inputVideoPath = p.join(projectRoot,_extractVideoId(videoUrl) ),
        originalMp3 = p.join(projectRoot, "${_extractVideoId(videoUrl)}.mp3"),
        imageCover = p.join(projectRoot, "cover.jpg"),
        taskId = DateTime.now().millisecondsSinceEpoch.toString() {
    // this.videoCosUrl = initialVideoCosUrl; // Example if passed
    // Or, if you have a Dart equivalent for settings.tx_cos_url and video.cosKey:
    // this.videoCosUrl = video != null ? '${YourDartSettings.txCosUrl}/${video.cosKey}' : null;
    _setupPaths();
  }

  static String _extractVideoId(String url) {
    final RegExp regExp = RegExp(
      r"(?:v=|\/)([0-9A-Za-z_-]{11})",
      caseSensitive: false,
      multiLine: false,
    );
    final Match? match = regExp.firstMatch(url);
    if (match != null && match.groupCount >= 1) {
      return match.group(1)!;
    }
    throw ArgumentError("Invalid YouTube URL: $url");
  }

  void _setupPaths() {
    if (videoId.isEmpty) return; // Or handle error

    // currentDir = p.join(projectRoot, videoId);

    // Ensure directories exist
    Directory(currentDir).createSync(recursive: true);
    Directory(p.join(currentDir, 'audio')).createSync(recursive: true);

    audioDir = p.join(currentDir, "audio");
    inputVideoPath = p.join(currentDir, "$videoId.mp4");
    outVideoPath = p.join(currentDir, "${videoId}_trans.mp4");
    translateMp4 = p.join(currentDir, "${videoId}_trans.mp4");
    originalMp3 = p.join(currentDir, "$videoId.mp3"); // Python used audio_dir here, but for consistency with video files, current_dir is used. Adjust if needed.
    // originalMp3 = p.join(audioDir!, "$videoId.mp3"); // Alternative based on Python structure
    originalSrt = p.join(currentDir, "en.srt");
    translateTxt = p.join(currentDir, "zh.txt");
    mindmapJson = p.join(currentDir, "$videoId.json");
    translateMd = p.join(currentDir, "$videoId.md");
    originalVtt = p.join(currentDir, "en.vtt");
    translateSrt = p.join(currentDir, "zh.srt");
    translateVtt = p.join(currentDir, "zh.vtt");
    imageCover = p.join(currentDir, "cover.jpg");
    okFile = p.join(currentDir, "ok");
    upVideoPath = inputVideoPath; // Default as per Python
  }

  String getRelativePath(String fullPath) {
    if (currentDir == null) {
      throw StateError("Paths not set up, currentDir is null.");
    }
    return p.relative(fullPath, from: currentDir!);
  }

  void setError(String message, [dynamic details]) {
    hasError = true;
    errorMessage = message;
    errorDetails = details;
    print("ERROR SET: $message - Details: $details");
  }

  void clearError() {
    hasError = false;
    errorMessage = null;
    errorDetails = null;
  }

  @override
  String toString() {
    return '''
TaskStateManager:
  Config:
    Video URL: $videoUrl
    Project Root: $projectRoot
    Video ID: $videoId
    Task ID: $taskId

  Paths:
    Current Dir: $currentDir
    Audio Dir: $audioDir
    Input Video Path: $inputVideoPath
    Output Video Path (Translated): $outVideoPath 
    Original MP3: $originalMp3
    Cover Image: $imageCover
    ... (add other paths as needed for debugging)

  State:
    Processed Video Path: $processedVideoPath 
    Extracted Frames Path: $extractedFramesPath
    Extracted Image Files: ${extractedImageFiles?.join(', ') ?? 'N/A'}
    Extracted Audio Path: $extractedAudioPath
    Extracted Audio File: $extractedAudioFile
    
  Error State:
    Has Error: $hasError
    Error Message: $errorMessage
    Error Details: $errorDetails
''';
  }
}

File: app/src/app/core/handler/task_state_manager.py
import os
from pathlib import Path
from app.core.logger import get_logger
from app.utils.time_utils import getTimeStamp
from app.utils.file_utils import create_and_write_file_pathlib
from app.models.VideoData import VideoData, MQTTMessageBuilder

logger = get_logger(__name__)

# 状态管理器
class TaskStateManager:
    def __init__(self, project_root, video=None, mqtt_service=None,session = None):
        self.video = video
        self.session = session
        self.id = video.id if video else None
        self.video_id = video.videoId if video else None
        self.video_url = video.url if video else None
        self.project_root = project_root
        self.task_id = getTimeStamp()
        self.mqtt_service = mqtt_service  # 添加 MQTT 服务引用

        self.state = {}
        self.cache = {}  # 内存缓存
        
        self.current_dir = None
        self.audio_dir = None
        self.input_video_path = None
        self.out_video_path = None
        self.translate_mp4 = None
        self.original_mp3 = None
        self.original_srt = None
        self.original_vtt = None
        self.translate_srt = None
        self.translate_txt = None
        self.translate_vtt = None
        self.image_cover = None
        self.up_video_path = None


        
        # 如果初始化时已提供 video，则设置相关路径
        if self.video_id:
            self._setup_paths()

    def _setup_paths(self):
        """设置与视频相关的所有路径"""
        self.current_dir = os.path.join(Path(self.project_root), self.video_id)
        
        # 确保目录存在
        os.makedirs(self.current_dir, exist_ok=True)
        os.makedirs(f'{self.current_dir}/audio', exist_ok=True)
        
        self.audio_dir = f"{self.current_dir}/audio"
        self.input_video_path = f"{self.current_dir}/{self.video_id}.mp4"
        self.out_video_path = f"{self.current_dir}/{self.video_id}_trans.mp4"
        self.translate_mp4 = f"{self.current_dir}/{self.video_id}_trans.mp4"
        self.original_mp3 = f"{self.current_dir}/{self.video_id}.mp3"
        self.original_srt = f"{self.current_dir}/en.srt"
        self.translate_txt = f"{self.current_dir}/zh.txt"
        self.mindmap_json = f"{self.current_dir}/{self.video_id}.json"
        self.translate_md = f"{self.current_dir}/{self.video_id}.md"
        self.original_vtt = f"{self.current_dir}/en.vtt"
        self.translate_srt = f"{self.current_dir}/zh.srt"
        self.translate_vtt = f"{self.current_dir}/zh.vtt"
        self.image_cover = f"{self.current_dir}/cover.jpg"
        self.ok_file = f"{self.current_dir}/ok"
        self.up_video_path = self.input_video_path

    def get_relative_path(self,pathName):
        """获取相对路径"""
        return os.path.relpath(self.current_dir, pathName)

    def set_video(self, video: VideoData):
        """动态设置 video 并更新相关路径"""
        self.video = video
        self.video_id = video.videoId
        self._setup_paths()
        
    def set_context(self, key, value):
        """设置上下文数据"""
        self.set_shared_data(key, value)

    def set_shared_data(self, key, value):
        """
        Set data to be shared between tasks.
        """
        self.state[key] = value

    def get_shared_data(self, key, default=None):
        """
        Get shared data between tasks.
        """
        return self.state.get(key, default)

    def get_file_name(self, input: str):
        return input.replace(self.project_root, "")

    def write_ok_file(self):
        create_and_write_file_pathlib(self.current_dir, "ok", "ok")


    def update_video_status(self,pamas):
        pass

    def send_mqtt_message(self, message):
        """通过 MQTT 服务发送消息"""
        if self.mqtt_service:
            return self.mqtt_service.publish(message)
        else:
            logger.error("MQTT 服务未初始化，无法发送消息")
            return None
